compiler/rustc_codegen_llvm/src/va_arg.rs

```diff
@@ -278,6 +278,177 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
+    // `VaArgSafe` trait. Currently, only pointers, f64, i32, u32, i64 and u64 implement this trait.
+
+    // typedef struct __va_list_tag {
+    //     unsigned int gp_offset;
+    //     unsigned int fp_offset;
+    //     void *overflow_arg_area;
+    //     void *reg_save_area;
+    // } va_list[1];
+    let va_list_addr = list.immediate();
+
+    let unsigned_int_offset = 4;
+    let ptr_offset = 8;
+    let gp_offset_ptr = va_list_addr;
+    let fp_offset_ptr = bx.inbounds_ptradd(va_list_addr, bx.cx.const_usize(unsigned_int_offset));
+
+    let layout = bx.cx.layout_of(target_ty);
+
+    // AMD64-ABI 3.5.7p5: Step 1. Determine whether type may be passed
+    // in the registers. If not go to step 7.
+
+    // AMD64-ABI 3.5.7p5: Step 2. Compute num_gp to hold the number of
+    // general purpose registers needed to pass type and num_fp to hold
+    // the number of floating point registers needed.
+
+    let mut num_gp_registers = 0;
+    let mut num_fp_registers = 0;
+
+    match layout.layout.backend_repr() {
+        BackendRepr::Scalar(scalar) => match scalar.primitive() {
+            rustc_abi::Primitive::Int(integer, _is_signed) => {
+                num_gp_registers += integer.size().bytes().div_ceil(8) as i32;
+            }
+            rustc_abi::Primitive::Float(float) => {
+                num_fp_registers += float.size().bytes().div_ceil(16) as i32;
+            }
+            rustc_abi::Primitive::Pointer(_) => {
+                num_gp_registers += 1;
+            }
+        },
+        BackendRepr::ScalarPair(..) => {
+            unreachable!("VaArgSafe is not implemented for scalar pairs")
+        }
+        BackendRepr::SimdVector { .. } => {
+            unreachable!("VaArgSafe is not implemented for SIMD vectors")
+        }
+        BackendRepr::Memory { .. } => { /* do nothing */ }
+    };
+
+    if num_gp_registers == 0 && num_fp_registers == 0 {
+        unreachable!("VaArgSafe is not implemented for ZSTs")
+    }
+
+    // AMD64-ABI 3.5.7p5: Step 3. Verify whether arguments fit into
+    // registers. In the case: l->gp_offset > 48 - num_gp * 8 or
+    // l->fp_offset > 176 - num_fp * 16 go to step 7.
+
+    let gp_offset_v = bx.load(bx.type_i32(), gp_offset_ptr, Align::from_bytes(8).unwrap());
+    let fp_offset_v = bx.load(bx.type_i32(), fp_offset_ptr, Align::from_bytes(4).unwrap());
+
+    let mut use_regs = bx.const_bool(false);
+
+    if num_gp_registers > 0 {
+        let max_offset_val = 48i32 - num_gp_registers * 8;
+        let fits_in_gp = bx.icmp(IntPredicate::IntULE, gp_offset_v, bx.const_i32(max_offset_val));
+        use_regs = fits_in_gp;
+    }
+
+    if num_fp_registers > 0 {
+        let max_offset_val = 176i32 - num_fp_registers * 16;
+        let fits_in_fp = bx.icmp(IntPredicate::IntULE, fp_offset_v, bx.const_i32(max_offset_val));
+        use_regs = if num_gp_registers > 0 { bx.and(use_regs, fits_in_fp) } else { fits_in_fp };
+    }
```

Comment:
The way this code is written is a bit puzzling to me, as it interleaves GP and FP types. Is this because clang does it this way, so that it can handle structs?

Reply:
effectively, yes. I'm not sure how much we should future-proof this, but just for the current state these could be separated

Reply:
also it's how the spec is written, but that has the same reason I'd guess